azure-functions-powershell-library/WorkerIndexingHelper.cs

```diff
@@ -0,0 +1,197 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using AzureFunctions.PowerShell.SDK;
+using System.Collections.ObjectModel;
+using System.Management.Automation.Language;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker
+{
+    internal class WorkerIndexingHelper
+    {
+        internal static List<FunctionInformation> IndexFunctions(string baseDir)
+        {
+            if (!Directory.Exists(baseDir))
+            {
+                throw new FileNotFoundException();
+            }
+            List<FileInfo> powerShellFiles = GetPowerShellFiles(Directory.CreateDirectory(baseDir));
+            List<FunctionInformation> rpcFunctionMetadatas = new List<FunctionInformation>();
+
+            foreach (FileInfo powerShellFile in powerShellFiles)
+            {
+                rpcFunctionMetadatas.AddRange(IndexFunctionsInFile(powerShellFile));
+            }
+
+            return rpcFunctionMetadatas;
+        }
+
+        private static IEnumerable<FunctionInformation> IndexFunctionsInFile(FileInfo powerShellFile)
+        {
+            List<FunctionInformation> fileFunctions = new List<FunctionInformation>();
+            var fileAst = Parser.ParseFile(powerShellFile.FullName, out _, out ParseError[] errors);
+            if (errors.Any())
+            {
+                throw new Exception($"Couldn't parse the file: {powerShellFile.FullName}");
+                // TODO: Probably don't throw here?
+                //return fileFunctions;
```

Comment:
I think it's fine to throw here, though I'll defer to the PowerShell SMEs for guidance on that.
My main concern here is the error message. Can we help the user diagnose why we couldn't parse their file? Perhaps we can link them to docs, or simply list out the formatting expectations here.

Reply:
Yeah - The reason I would suggest not throwing is because even if there is a poorly-formatted PowerShell file in the function app directory, we may want to parse other, valid files for functions and just proceed anyway. @AnatoliB @Francisco-Gamino any opinions one way or another? If we decide to throw, I'll improve the error message

Reply:
@andystaples - I'll leave it to the other SMEs to make the final decision.

However, my 2 cents is that, if there's some poorly-formatted PowerShell file, it would be best to "fail early" than to have users deploy a (potentially) partially-working application.

Reply:
Agreed. In the latest commit, you can see a test for a poorly-formatted ps1 file and the exception it should throw

Reply:
We settled on NonTerminatingErrors for this case. I need to update the code and the test cases to reflect this.

Reply:
Done